### PR TITLE
Adding UT coverage for VM validation in VRG

### PR DESCRIPTION
**Pull request by abhijeet219**
UT coverage for VM-DR epic:
Writing unit tests and validations of VRG status for different combination of VMs(with and without protection label) and their states.

Jira story [#RHSTOR-7348](https://issues.redhat.com/browse/RHSTOR-7348)

### CODE REVIEW

File: internal/controller/vrg_volrep_test.go [hunk 1]
```diff
@@ -1057,6 +1061,426 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			Expect(condition).NotTo(BeNil())
+			Expect(condition.Status).To(Equal(metav1.ConditionFalse))
+			Expect(condition.Reason).To(Equal("ClusterDataConflictSecondary"))
+
+			deleteVM(vm, vmNamespacedName)
+
+			vrgTestCase.cleanupVRG()
+		})
+
+		It("VM do not have the protection label, but is in protected vms list of secondary vrg", func() {
```
[the remark this answers is not in the document]
Comment (pruthvitd) [reply internal/controller/vrg_volrep_test.go:1228]: Hi @nirs, these tests are added specifically to test VM-recipe. The virtualmachine instances gets validated by the VRG controller, in-order to simulate the scenarios we need to create VM instances in testenv.

File: hack/test/virtualmachines.kubevirt.io.yaml
Comment (nirs) [inline hack/test/virtualmachines.kubevirt.io.yaml:1]: The minimized version looks nice. Please add a comment explaining that this is minimized version of the original crd, and add a link to source you took it from.

File: go.mod [hunk 2]
```diff
@@ -35,7 +35,7 @@ require (
 	k8s.io/component-base v0.31.1
 	k8s.io/kube-openapi v0.0.0-20241105132330-32ad38e42d3f
 	k8s.io/utils v0.0.0-20241104100929-3ea5e8cea738
-	kubevirt.io/api v1.5.0
+	kubevirt.io/api v1.5.1
```
Comment (nirs) [inline go.mod]: Upgrading to 1.5.1 is not related to the test change, this should be done separately, and use latest version:
https://github.com/kubevirt/kubevirt/releases/tag/v1.5.2

Note that in drenv we are still testing kubevirt 1.3. We need to upgrade it to 1.5.2 as well.

File: internal/controller/vrg_volrep_test.go [hunk 3]
```diff
@@ -1072,6 +1076,426 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
 		})
 	})
 
+	var vrgTestCase *vrgTest
+	var vrgTestCase1 *vrgTest
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: These names do not describe the purpose of the variable. It would be better to define them at the context scope so we use the same short variable name for the entire context.

Looking at the new code vrgTestCase1 is not used in this context.

File: internal/controller/vrg_volrep_test.go [hunk 4]
```diff
@@ -1072,6 +1076,426 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			vrg := vrgTestCase.getVRG()
+
+			condition := meta.FindStatusCondition(vrg.Status.Conditions, "DataReady")
+			Expect(condition).NotTo(BeNil())
+			Expect(condition.Status).To(Equal(metav1.ConditionFalse))
+			Expect(condition.Reason).To(Equal("Error"))
+			Expect(condition.Message).To(Equal("VolumeReplicationGroup is not in the admin namespace: " +
+				"VolumeReplicationGroup is not allowed to protect namespaces"))
+
+			vrgTestCase.cleanupVRG()
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: Is this called if an assertion failed before the test?

for example
https://github.com/RamenDR/ramen/pull/2052/commits/39c2d56a615959116e3d3991051a2bc10619f42b#diff-af799cf0d298b89e62ed4257879ffdf53d91e60f4f4871f01bc3811c61366209R1113

Cleanup code must be called from cleanup handler so it is always called, even on failures.

File: internal/controller/vrg_volrep_test.go [hunk 5]
```diff
@@ -1072,6 +1076,426 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+
+			// After removing the label from VM, the condition should be True
+			// It will not be able to identify VM by name due to name conflict
+			Eventually(func() bool {
+				return (meta.FindStatusCondition((vrgTestCase.getVRG()).Status.Conditions,
+					"NoClusterDataConflict")).Status == metav1.ConditionTrue
+			}, vrgtimeout, vrginterval).Should(BeTrue(), "while waiting for VRG TRUE condition on secondary %s",
+				vrgTestCase.vrgName)
+
+			deleteVM(vm, vmNamespacedName)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: Other tests do not delete the vm - should they?

File: internal/controller/vrg_volrep_test.go [hunk 6]
```diff
@@ -1072,6 +1076,426 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+
+			// Even after removing the protection label, the condition should be False
+			// because the protected VMs in the vrg Spec still exists
+			vrg = vrgTestCase.getVRG()
+			condition = meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+			Expect(condition).NotTo(BeNil())
+			Expect(condition.Status).To(Equal(metav1.ConditionFalse))
+			Expect(condition.Reason).To(Equal("ClusterDataConflictSecondary"))
+
+			deleteVM(vm1, vmNamespacedName1)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:1482]: This will not be called if the assetions before fail, but we don't call this later in a cleanup.

File: hack/test/virtualmachines.kubevirt.io.yaml
Comment (abhijeet219) [reply hack/test/virtualmachines.kubevirt.io.yaml:1]: Done

File: internal/controller/vrg_volrep_test.go (hunk 3 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: Addressed. I think the confusion is gone now

File: internal/controller/vrg_volrep_test.go (hunk 4 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: Tried to address this using DeferCleanup() calls for vm and vrg resources

File: internal/controller/vrg_volrep_test.go (hunk 5 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: they are deleting now. Thanks

File: internal/controller/vrg_volrep_test.go (hunk 6 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go:1482]: Addressed using the DeferCleanup() calls for resources

File: internal/controller/vrg_volrep_test.go [hunk 7]
```diff
@@ -271,7 +273,7 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
 		It("populates the S3 store with PVs/PVCs and start vrg as primary to check that the PVs/PVCs are restored", func() {
 			restoreTestTemplate.s3Profiles = []string{s3Profiles[vrgS3ProfileNumber].S3ProfileName}
 			numPVs := 3
-			vtest := newVRGTestCaseCreate(0, restoreTestTemplate, true, false)
+			vtest := newVRGTestCaseCreate(0, restoreTestTemplate, true, false, "", "primary", nil)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: This is very hard to follow. We already have too many arguments which are hard to follow (what does true, false mean?). This adds 3 more arguments.

The right way is to create an Options type and convert the call to something like:

    newVRGTestCaseCreate(testOptions{Key: value, ...})

File: internal/controller/vrg_volrep_test.go [hunk 8]
```diff
@@ -1081,6 +1085,419 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
 		})
 	})
 
+	var vrgTestCase *vrgTest
+	Context("Create VRG and validate VM status", func() {
+		BeforeEach(func() {
+			ramenConfig.RamenOpsNamespace = "ramen-ops"
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:1085]: This should always be set in ramenconfig. Why do we test without this?

File: internal/controller/vrg_volrep_test.go [hunk 9]
```diff
@@ -1081,6 +1085,419 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+		vrcLabels := genVRCLabels(replicationIDs[0], storageID, "ramen")
+		vrgEmptySCTemplate := &template{
+			ClaimBindInfo:          corev1.ClaimBound,
+			VolumeBindInfo:         corev1.VolumeBound,
+			schedulingInterval:     "1h",
+			replicationClassName:   "test-replicationclass",
+			vrcProvisioner:         "manual.storage.com",
+			scProvisioner:          "manual.storage.com",
+			storageIDLabels:        storageIDLabel,
+			replicationClassLabels: vrcLabels,
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: Anything here which cannot be a default value?

File: internal/controller/vrg_volrep_test.go [hunk 10]
```diff
@@ -1081,6 +1085,419 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+		}
+
+		It("sets up PVCs, PVs and VRGs - with nil/empty StorageClassName", func() {
+			vrgEmptySCTemplate.s3Profiles = []string{s3Profiles[vrgS3ProfileNumber].S3ProfileName}
+		})
+
+		// Scenarios when VM is not there
+		It("creates VRG in non-admin namespace for primary", func() {
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: We don't have tests for discovered using this configuration?

File: internal/controller/vrg_volrep_test.go [hunk 11]
```diff
@@ -1081,6 +1085,419 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+		// Scenarios where a dummy VM is created
+		It("VM has protection label and is in protected vms list of primary vrg", func() {
+			createVM("vm-1-test")
+			protectedVMs := []string{"vm-1-test"}
+			vmNamespacedName := types.NamespacedName{
+				Name:      "vm-1-test",
+				Namespace: "default",
+			}
+			vm := getVM(vmNamespacedName)
+			DeferCleanup(deleteVM, vm, vmNamespacedName)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:1117]: This should be after the line creating the vm (line 11174). When using standard go test the best way to ensure cleanup is to use:

```go
createVM(t *testing.T, ...) {
    // create the vm here...
    
    t.Cleanup(func() {
         // remove vm here...
    })
}
```

I'm not sure what is the ginkgo equivalent but it should support the same.

Creating the vm and depending on getVM to return it to do the cleanup is wrong and will cause leftovers vms which will fail the next tests or worse may cause them to succeed when the code is broken.

File: internal/controller/vrg_volrep_test.go [hunk 12]
```diff
@@ -1081,6 +1085,419 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			protectVM(vm)
+
+			vrgTestCase = newVRGTestCaseCreateAndStart(1, vrgEmptySCTemplate, true, false, true,
+				ramenConfig.RamenOpsNamespace, "primary", protectedVMs)
+			DeferCleanup(vrgTestCase.cleanupVRG)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: This looks correct since we handle cleanup right after creating the test case, But this means we have to remember to call this function. It will be better if the vrg cleanup is part of the code creating it, cleaning up when the current test is done.

File: internal/controller/vrg_volrep_test.go [hunk 13]
```diff
@@ -1802,18 +2246,169 @@ func (v *vrgTest) VRGTestCaseStart() {
 	}
 }
 
+func createVM(vmName string) {
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:2313]: The implementation is actually "create if needed" or "ensure vm". Why are we ok with trying to creating an existing vm? It seems that we hide missing cleanup at the end of the test. This can cause tests to use the wrong vm.

For example, previous test created a vm with different confguration, and current tests will use the wrong vm configuration which can fail or hide errors in the code.

File: internal/controller/vrg_volrep_test.go [hunk 14]
```diff
@@ -1802,18 +2246,169 @@ func (v *vrgTest) VRGTestCaseStart() {
...
+	return vm
+}
+
+func deleteVM(vm *virtv1.VirtualMachine, vmNamespacedName types.NamespacedName) {
+	By("deleting VM " + vm.Name)
+
+	err := k8sClient.Delete(context.TODO(), vm)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		Expect(err).To(BeNil(), "failed to delete VM %s", vm.Name)
+	}
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:2347]: If we get IsNotFound error we can return, no ned to wait for deletion since the vm was already delete.

But this check looks wrong - we enter this block if err != nil, and then we validate the error is not nil which we already know. So we should fail the test with a message, no need of the Expect() boilerplate. Maybe the entire if block can be converted to Expect() for err == nil or !k8serrors.IsNotFound(err). If

File: internal/controller/vrg_volrep_test.go [hunk 15]
```diff
@@ -1802,18 +2246,169 @@ func (v *vrgTest) VRGTestCaseStart() {
...
+		return k8sClient.Get(context.TODO(), vmNamespacedName, vm)
+	}, vrgtimeout, vrginterval).Should(MatchError(k8serrors.NewNotFound(schema.GroupResource{
+		Group:    virtv1.SchemeGroupVersion.Group,
+		Resource: "VirtualMachine",
+	}, vm.Name)), "VM %s was not deleted", vm.Name)
+}
+
+func generateVM(vmName, claimName, namespace string) *virtv1.VirtualMachine {
+	return vm(vmName, claimName, namespace)
+}
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: What it the purpose of this helper? it just call vm().

File: internal/controller/vrg_volrep_test.go [hunk 16]
```diff
@@ -1802,18 +2246,169 @@ func (v *vrgTest) VRGTestCaseStart() {
...
+
+// Add protected-by-ramendr label to VM
+func protectVM(vm *virtv1.VirtualMachine) {
+	labels := vm.GetLabels()
+	labels["ramendr.openshift.io/k8s-resource-selector"] = "protected-by-ramendr"
+	vm.SetLabels(labels)
+	err := k8sClient.Update(context.TODO(), vm)
+	Expect(err).To(BeNil(), "failed to update VM %s", vm.Name)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: This can fail with a conflict. ALL updates should use retry.RetryOnConflict helper we already use in many places. Example from e2e:
https://github.com/RamenDR/ramen/blob/eebc5c0cb46af2eea145e7d40feef09681f6b110/e2e/dractions/retry.go#L155

File: internal/controller/vrg_volrep_test.go [hunk 17]
```diff
@@ -1802,18 +2246,169 @@ func (v *vrgTest) VRGTestCaseStart() {
...
+	labels := vm.GetLabels()
+	delete(labels, "ramendr.openshift.io/k8s-resource-selector")
+	vm.SetLabels(labels)
+	err := k8sClient.Update(context.TODO(), vm)
+	Expect(err).To(BeNil(), "failed to update VM %s", vm.Name)
+}
+
+// nolint:funlen
+func vm(vmName, claimName, vmNamespaceName string) *virtv1.VirtualMachine {
+	vmTemplate := &virtv1.VirtualMachine{
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: Why do we call this vmTemplate and not vm?

File: internal/controller/vrg_volrep_test.go [hunk 18]
```diff
@@ -1802,18 +2246,169 @@ func (v *vrgTest) VRGTestCaseStart() {
...
+func unprotectVM(vm *virtv1.VirtualMachine) {
+	labels := vm.GetLabels()
+	delete(labels, "ramendr.openshift.io/k8s-resource-selector")
+	vm.SetLabels(labels)
+	err := k8sClient.Update(context.TODO(), vm)
+	Expect(err).To(BeNil(), "failed to update VM %s", vm.Name)
+}
+
+// nolint:funlen
+func vm(vmName, claimName, vmNamespaceName string) *virtv1.VirtualMachine {
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: This function name is too short and confict with local variables. Maybe newVM()?  Check how we call helpers to create VRG and PVC and use similar name.

File: internal/controller/vrg_volrep_test.go [hunk 19]
```diff
@@ -2123,6 +2718,61 @@ func (v *vrgTest) createVRG() {
 		"failed to create VRG %s in %s", v.vrgName, v.namespace)
 }
 
+func (v *vrgTest) createVRGwithVMProtection() {
+	By("creating VRG " + v.vrgName)
+
+	schedulingInterval := "1h"
+	replicationClassLabels := map[string]string{"protection": "ramen"}
+
+	vrg := &ramendrv1alpha1.VolumeReplicationGroup{
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:2772]: What is the difference between VRG with and without vm protection? Can we use a VRG created by existing code, and add the few missing details? Here we duplicated lot of code which the vm tests should not care about and someone will have to maintain all the duplicate code.

This code is also hard to understand. Developer reading this code cannot learn from it what is needed for protecting vms. If the code was using another function the create the vrg and add the protection bits, it would be very clear what are these bits.

File: internal/controller/vrg_volrep_test.go [hunk 20]
```diff
@@ -2123,6 +2718,61 @@ func (v *vrgTest) createVRG() {
...
+
+	err := k8sClient.Create(context.TODO(), vrg)
+	expectedErr := k8serrors.NewAlreadyExists(
+		schema.GroupResource{
+			Group:    "ramendr.openshift.io",
+			Resource: "volumereplicationgroups",
+		},
+		v.vrgName)
+	Expect(err).To(SatisfyAny(BeNil(), MatchError(expectedErr)),
+		"failed to create VRG %s in %s", v.vrgName, v.namespace)
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:2792]: This function mix:
- creating vrg object
- setting vm protection bits
- creating a resource

This is too much for one function.

File: internal/controller/vrg_volrep_test.go [hunk 21]
```diff
@@ -2312,6 +2962,7 @@ func updateVRG(desired *ramendrv1alpha1.VolumeReplicationGroup) {
 		current := vrgGet(client.ObjectKeyFromObject(desired))
 
 		current.Spec = desired.Spec
+		current.ObjectMeta = desired.ObjectMeta
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go]: Why do need to set ObjectMeta now and it was not required before? Do we have this issue in other tests?

This worth its own commit with a good message explaining the issue.

File: internal/controller/vrg_volrep_test.go (hunk 8 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go:1085]: This one is for discovered apps, hence only applicable to the newly added test-cases. I'll make it better by putting the hard-coded value into a const variable.

File: internal/controller/vrg_volrep_test.go (hunk 10 above)
Comment (pruthvitd) [reply internal/controller/vrg_volrep_test.go]: Yes, there are no UT tests added for discovered apps.

File: internal/controller/vrg_volrep_test.go (hunk 7 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: Reverted it like original function.

File: internal/controller/vrg_volrep_test.go [hunk 22]
```diff
@@ -1081,6 +1083,567 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			vrgTestCase.protectedVMs = protectedVMs
+			vrgTestCase.VRGTestCaseStart()
+			DeferCleanup(vrgTestCase.cleanupVRG)
+			vrgTestCase.verifyVRGStatusExpectation(true, "Unused")
+
+			vrg = vrgTestCase.getVRG()
+			condition := meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+			Expect(condition).NotTo(BeNil())
+			Expect(condition.Status).To(Equal(metav1.ConditionTrue))
+			Expect(condition.Reason).To(Equal("NoConflictDetected"))
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: Lets replace Expect with Eventually()

Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: > Lets replace Expect with Eventually()

Corrected this in three test-cases, where I was not doing validation using Eventually() block, or using an Eventually() block before actual validation.
Thanks for pointing out!

File: internal/controller/vrg_volrep_test.go [hunk 23]
```diff
@@ -1081,6 +1083,576 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+					return condition.Status
+				}
+
+				return metav1.ConditionUnknown
+			}, vrgtimeout, vrginterval).Should(
+				Equal(metav1.ConditionFalse), "while waiting for VM conflict validation to complete")
+
+			vrg = vrgTestCase.getVRG()
+			condition := meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+			Expect(condition).NotTo(BeNil())
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: Simplify this by including them inside the Eventually block. Also avoid unnecessary get call at line#1173

**Review by pruthvitd — CHANGES_REQUESTED** [review]
Please dont combine, main code changes with UT enhancements.

File: internal/controller/volumereplicationgroup_controller.go [hunk 24]
```diff
@@ -2440,10 +2441,14 @@ func (v *VRGInstance) CheckForVMNameConflictOnSecondary(vmNamespaceList, vmList
 		return nil
 	}
 
-	v.log.Info(fmt.Sprintf("found conflicting VM[%v] on secondary", foundVMs))
+	if len(foundVMs) > 0 {
```
Comment (pruthvitd) [inline internal/controller/volumereplicationgroup_controller.go]: This if is redundant here.

File: internal/controller/vrg_volrep_test.go (hunk 9 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: I've implemented this with a constructor function approach that provides defaults for the 3 repetitive fields (schedulingInterval, vrcProvisioner, scProvisioner) while keeping other fields flexible.

File: internal/controller/vrg_volrep_test.go [hunk 25]
```diff
@@ -1081,22 +1079,593 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+				vrg = vrgTestCase.getVRG()
+				condition := meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+
+				return condition != nil &&
+					condition.Status == metav1.ConditionFalse &&
+					condition.Reason == "ClusterDataConflictSecondary"
+			}, vrgtimeout, vrginterval).Should(BeTrue())
+		})
+
+		It("Secondary VRG: VM without protection label in VRG spec should trigger validation and report conflict", func() {
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: The **Secondary** VRG should not report conflict if VM not matching protection label mentioned in VRG.

File: internal/controller/vrg_volrep_test.go [hunk 26]
```diff
@@ -1081,22 +1079,593 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+				vrg = vrgTestCase.getVRG()
+				condition := meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+
+				return condition != nil &&
+					condition.Status == metav1.ConditionTrue &&
+					condition.Reason == "NoConflictDetected"
+			}, vrgtimeout, vrginterval).Should(BeTrue())
+		})
+
+		It("Primary VRG: Name of VM existing differs from VRG spec results in conflict", func() {
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: Can we rephrase like this:
"Primary VRG: Conflict if protected namespace VMs don’t match VRG spec names."

File: internal/controller/vrg_volrep_test.go [hunk 27]
```diff
@@ -1081,22 +1079,593 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			vrgTestCase = newVRGTestCaseCreate(1, vrgEmptySCTemplate, true, false)
+			vrgTestCase.namespace = ramenConfig.RamenOpsNamespace
+			vrgTestCase.replicationState = "secondary"
+			vrgTestCase.protectedVMs = []string{"vm-1-test"}
+			vrgTestCase.pvcLabels["ramendr.openshift.io/k8s-resource-selector"] = "protected-by-ramendr"
+			vrgTestCase.VRGTestCaseStart()
+			DeferCleanup(vrgTestCase.cleanupVRGAndPVCs)
+
+			pvcName := vrgTestCase.pvcNames[0].Name
+			createVMWithPVC("vm-1-test", pvcName)
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go:1235]: Creating conflicting VM on secondary with the name, same as protected VMName from VRG spec.

File: internal/controller/vrg_volrep_test.go [hunk 28]
```diff
@@ -1081,22 +1079,593 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			}
+			vm := getVM(vmNamespacedName)
+			DeferCleanup(deleteVM, vm, vmNamespacedName)
+
+			vrgTestCase = newVRGTestCaseCreate(1, vrgEmptySCTemplate, true, false)
+			vrgTestCase.namespace = ramenConfig.RamenOpsNamespace
+			vrgTestCase.protectedVMs = protectedVMs
+			vrgTestCase.VRGTestCaseStart()
+			DeferCleanup(vrgTestCase.cleanupVRGAndPVCs)
+			vrgTestCase.verifyVRGStatusExpectation(true, "Unused")
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: DataReady Condition will not be unused for Primary VRG.

File: internal/controller/vrg_volrep_test.go [hunk 29]
```diff
@@ -1081,22 +1079,593 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+				if vrg == nil {
+					return false
+				}
+				condition := meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+
+				return condition != nil && condition.Status == metav1.ConditionTrue
+			}, vrgtimeout, vrginterval).Should(BeTrue())
+		})
+
+		It("Primary VRG-Secondary VRG: After VM deletion, primary shows conflict, secondary shows no-conflict", func() {
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: We can drop this scenario, as its already covered in previous tests.

**Review by pruthvitd — CHANGES_REQUESTED** [review]

File: internal/controller/vrg_volrep_test.go (hunk 28 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: `VRGConditionReasonUnused` is returned when `len(v.volRepPVCs) == 0`, and here in VM validation scenarios the VRG selector doesn't find any matching PVCs. So unused is the right thing to use, since `len(v.volRepPVCs) == 0` will be true.
In case of VM validation scenarios, this will be `VRGConditionReasonUnused` for both Primary and Secondary.

File: internal/controller/vrg_volrep_test.go (hunk 25 above)
Comment (abhijeet219) [reply internal/controller/vrg_volrep_test.go]: Secondary VRG will report if VM in the protectedVMs list of VRG spec exists, or VM with any name has the protection label. So this is a correct scenario.
Updated the testcase description better to understand. Thanks

File: internal/controller/vrg_volrep_test.go [hunk 30]
```diff
@@ -1081,22 +1079,524 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+
+			protectVM(vm1)
+			protectVM(vm2)
+
+			vrgTestCase = newVRGTestCaseCreate(1, vrgEmptySCTemplate, true, false)
+			vrgTestCase.namespace = ramenConfig.RamenOpsNamespace
+			vrgTestCase.protectedVMs = []string{"vm-1-test", "vm-2-test"}
+			vrgTestCase.VRGTestCaseStart()
+			DeferCleanup(vrgTestCase.cleanupVRGAndPVCs)
+			vrgTestCase.verifyVRGStatusExpectation(true, "Unused")
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: Like mentioned in the earlier comment, this check can be removed.

File: internal/controller/vrg_volrep_test.go [hunk 31]
```diff
@@ -1081,22 +1079,524 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
...
+			unprotectVM(vm2)
+
+			Eventually(func() bool {
+				vrg := vrgTestCase.getVRG()
+				if vrg == nil {
+					return false
+				}
+				condition := meta.FindStatusCondition(vrg.Status.Conditions, "NoClusterDataConflict")
+
+				return condition != nil && condition.Status == metav1.ConditionFalse
```
Comment (pruthvitd) [inline internal/controller/vrg_volrep_test.go]: It would be beneficial to include an additional condition check for condition.Reason == "ClusterDataConflictPrimary" as part of this validation.

File: internal/controller/vrg_volrep_test.go (hunk 28 above)
Comment (pruthvitd) [reply internal/controller/vrg_volrep_test.go]: Thanks for the clarification — that makes sense in the context of VM validation scenarios where there are no PVC associated with the VM.
However, the intention of this test is to simulate a realistic DR scenario, and in such cases, a Primary VRG would typically have associated PVCs and reflect a Ready condition. Using Unused here may not align with the expected behavior in production-like environments and could lead to confusion when interpreting test results.
Also, as noted, the Eventually block already verifies that VRG.Status.Conditions are populated, so this specific check seems redundant and can be safely removed without impacting the test's purpose.
Let me know if you'd prefer to adjust the test setup to include PVCs for a more representative validation -- which you might already be addressing in the PVC conflict PR. For this PR, we can proceed with removing the check.

**Review by pruthvitd — APPROVED** [review]
Looking good, thanks.
